src/output-json.c

```diff
@@ -838,11 +838,21 @@ JsonBuilder *CreateEveHeader(const Packet *p, enum OutputJsonLogDirection dir,
         JsonAddrInfoInit(p, dir, &addr_info);
         addr = &addr_info;
     }
-    jb_set_string(js, "src_ip", addr->src_ip);
-    jb_set_uint(js, "src_port", addr->sp);
-    jb_set_string(js, "dest_ip", addr->dst_ip);
-    jb_set_uint(js, "dest_port", addr->dp);
-    jb_set_string(js, "proto", addr->proto);
+    if (strlen(addr->src_ip) > 0) {
+        jb_set_string(js, "src_ip", addr->src_ip);
+    }
+    if (addr->sp > 0) {
```

Comment:
0 is a valid port

Reply:
see also https://redmine.openinfosecfoundation.org/issues/6860

Reply:
Interesting.

Considering the issues with ICMP, I propose the following approach:
Instead of checking whether the ipproto is TCP/UDP/SCTP every time it's needed (to determine if a port can be logged), we set a flag that indicates when ports are elegible to be logged.

```
diff --git a/src/output-json.c b/src/output-json.c
index b0c1b7405..770bb3160 100644
--- a/src/output-json.c
+++ b/src/output-json.c
@@ -575,8 +575,10 @@ void JsonAddrInfoInit(const Packet *p, enum OutputJsonLogDirection dir, JsonAddr
         case IPPROTO_SCTP:
             addr->sp = sp;
             addr->dp = dp;
+            addr->log_port = true;
             break;
         default:
+            addr->log_port = false;
             break;
     }

@@ -838,11 +840,21 @@ JsonBuilder *CreateEveHeader(const Packet *p, enum OutputJsonLogDirection dir,
         JsonAddrInfoInit(p, dir, &addr_info);
         addr = &addr_info;
     }
-    jb_set_string(js, "src_ip", addr->src_ip);
-    jb_set_uint(js, "src_port", addr->sp);
-    jb_set_string(js, "dest_ip", addr->dst_ip);
-    jb_set_uint(js, "dest_port", addr->dp);
-    jb_set_string(js, "proto", addr->proto);
+    if (strlen(addr->src_ip) > 0) {
+        jb_set_string(js, "src_ip", addr->src_ip);
+    }
+    if (addr->log_port > 0) {
+        jb_set_uint(js, "src_port", addr->sp);
+    }
+    if (strlen(addr->src_ip) > 0) {
+        jb_set_string(js, "dest_ip", addr->dst_ip);
+    }
+    if (addr->log_port > 0) {
+        jb_set_uint(js, "dest_port", addr->dp);
+    }
+    if (strlen(addr->proto) > 0) {
+        jb_set_string(js, "proto", addr->proto);
+    }

     /* icmp */
     switch (p->proto) {
diff --git a/src/output-json.h b/src/output-json.h
index 353585262..7b618e64b 100644
--- a/src/output-json.h
+++ b/src/output-json.h
@@ -50,6 +50,7 @@ typedef struct JsonAddrInfo_ {
     Port sp;
     Port dp;
     char proto[JSON_PROTO_LEN];
+    bool log_port;
 } JsonAddrInfo;
```

Reply:
I think this makes sense.